File: monolith-app/src/main/java/com/sso/monolith/entity/Product.java
package com.sso.monolith.entity;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.Instant;
import lombok.*;

/**
 * Thực thể Đại diện cho thông tin Sản phẩm (Product) trong cơ sở dữ liệu Monolith.
 *
 * <p>Ánh xạ với bảng `products` trong CSDL `monolith_db`.
 *
 * @author SSO Platform Team
 * @since Sprint 07
 */
@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @Column(nullable = false)
  private String name;

  @Column(columnDefinition = "TEXT")
  private String description;

  @Column(nullable = false, precision = 18, scale = 2)
  private BigDecimal price;

  @Column(nullable = false)
  private Integer stock;

  @Column(length = 100)
  private String category;

  @Column(name = "image_url", length = 500)
  private String imageUrl;

  @Builder.Default
  @Column(nullable = false)
  private Boolean active = true;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "created_by", nullable = false)
  private UserProfile createdBy;

  @Column(name = "created_at", nullable = false, updatable = false)
  private Instant createdAt;

  @Column(name = "updated_at", nullable = false)
  private Instant updatedAt;

  @PrePersist
  protected void onCreate() {
    Instant now = Instant.now();
    this.createdAt = now;
    this.updatedAt = now;
    if (this.active == null) {
      this.active = true;
    }
  }

  @PreUpdate
  protected void onUpdate() {
    this.updatedAt = Instant.now();
  }
}

File: monolith-frontend/src/types/product.ts
/**
 * Product-related API Request and Response Types.
 */

export interface Product {
  id: number;
  name: string;
  price: number;
  stock: number;
  description: string | null;
  imageUrl: string | null;
  category: string | null;
  active: boolean;
  createdAt: string;
  updatedAt: string;
  createdBy: string | null;
}

export interface CreateProductRequest {
  name: string;
  price: number;
  stock: number;
  description?: string;
  imageUrl?: string;
  category?: string;
}

export interface UpdateProductRequest {
  name: string;
  price: number;
  stock: number;
  description?: string;
  imageUrl?: string;
  category?: string;
  active: boolean;
}
